File: src/main/java/Generator.java
import BladeRunner.NPC;
import BladeRunner.Table.*;

import java.util.concurrent.ThreadLocalRandom;

public class Generator {
    public static String createCase() {
        StringBuffer emergingCase = new StringBuffer();

        // create case
        emergingCase.append("ASSIGNMENT: \n");
        emergingCase.append(AssignmentTables.generateAssignment() + "\n");

        // create NPCs
        emergingCase.append("NPCS: \n");
        int numberOfNPCsToCreate = (ThreadLocalRandom.current().nextInt(1, 3 + 1)+3);
        NPC[] npcs = NPCTables.generateNPCs(numberOfNPCsToCreate);
        for(int i=0; i<numberOfNPCsToCreate; i++){
            emergingCase.append(npcs[i].toString()+"\n");
        }

        // scene of the crime
        emergingCase.append("LOCATIONS: \n");
        for (int i=0; i<3; i++) {
            emergingCase.append(LocationTables.generateLocation() + "\n");
        }

        //clues
        emergingCase.append("CLUES: \n");
        for (int i=0; i<5; i++) {
            emergingCase.append(CluesTable.generateClue() + "\n");
        }

        //mood pieces
        emergingCase.append("MOODS: \n");
        for (int i=0; i<3; i++) {
            emergingCase.append(SupplementaryTables.createMoodPiece() + "\n");
        }

        //twist
        emergingCase.append("TWIST: \n");
        emergingCase.append(SupplementaryTables.createTwist() + "\n");

        //final confrontation
        emergingCase.append("FINAL CONFRONTATION: \n");
        emergingCase.append(SupplementaryTables.createTwist() + "\n");

        return emergingCase.toString();
    }
}

File: js/generator.js
import { generateAssignment } from './tables/assignments.js';
import { generateNpcs, formatNpc } from './tables/npcs.js';
import { generateLocation, formatLocation } from './tables/locations.js';
import { generateClue, formatClue } from './tables/clues.js';
import { createTwist, createMoodPiece, formatMoodPiece } from './tables/supplementary.js';

// Java: ThreadLocalRandom.nextInt(1, 3 + 1) + 3 -> 1..3, then +3 -> 4..6.
function randomNpcCount() {
  return Math.floor(Math.random() * 3) + 1 + 3;
}

export function generateCase() {
  return {
    assignment: generateAssignment(),
    npcs: generateNpcs(randomNpcCount()),
    locations: Array.from({ length: 3 }, () => generateLocation()),
    clues: Array.from({ length: 5 }, () => generateClue()),
    moods: Array.from({ length: 3 }, () => createMoodPiece()),
    twist: createTwist(),
    // BUG (ported from Generator.java): this should call createFinalConfrontation(),
    // but the original calls createTwist() again. Fixed in Step 2.
    finalConfrontation: createTwist(),
  };
}

export function formatCase(c) {
  const lines = [];
  lines.push('ASSIGNMENT:', c.assignment, '');
  lines.push('NPCS:');
  c.npcs.forEach((npc) => lines.push(formatNpc(npc)));
  lines.push('');
  lines.push('LOCATIONS:');
  c.locations.forEach((loc) => lines.push(formatLocation(loc)));
  lines.push('');
  lines.push('CLUES:');
  c.clues.forEach((clue) => lines.push(formatClue(clue)));
  lines.push('');
  lines.push('MOODS:');
  c.moods.forEach((mood) => lines.push(formatMoodPiece(mood)));
  lines.push('');
  lines.push('TWIST:', c.twist, '');
  lines.push('FINAL CONFRONTATION:', c.finalConfrontation);
  return lines.join('\n');
}
